code/game/machinery/frame.dm

```diff
@@ -0,0 +1,307 @@
+			for(var/I in req_components)
+				if(req_components[I] > 0)
+					component_list += "[num2text(req_components[I])] [req_component_names[I]]"
+			D = "Requires [english_list(component_list)]."
+		desc = D
+
+/obj/structure/frame/New(var/loc, var/dir, var/building = 0, var/obj/item/frame/frame_type, var/obj/item/frame/icon, mob/user as mob)
+	..()
+
+	if(building)
+		src.frame_type = frame_type
+		icon_state = "[frame_type]_0"
+
+		if(frame_type == "alarm" || frame_type == "display")
+			if(loc)
+				src.loc = loc
+
+			state = 0
+			switch(frame_type)
+				if("alarm")
+					src.icon = icon
+					if(dir)
+						src.set_dir(dir)
+					pixel_x = (dir & 3)? 0 : (dir == 4 ? -24 : 24)
+					pixel_y = (dir & 3)? (dir ==1 ? -24 : 24) : 0
+				if("display")
+					src.icon = icon
+					pixel_x = (dir & 3)? 0 : (dir == 4 ? -32 : 32)
+					pixel_y = (dir & 3)? (dir ==1 ? -32 : 32) : 0
+			update_icon()
+			return
+
+	if(frame_type == "computer" || frame_type == "machine")
+		density = 1
+
+/obj/structure/frame/attackby(obj/item/P as obj, mob/user as mob)
+	switch(state)
+		if(0)
+			if(istype(P, /obj/item/weapon/wrench))
+				playsound(src.loc, 'sound/items/Ratchet.ogg', 50, 1)
+				if(do_after(user, 20))
+					user << "<span class='notice'>You wrench the frame into place.</span>"
+					src.anchored = 1
+					src.state = 1
+			if(istype(P, /obj/item/weapon/weldingtool))
+				var/obj/item/weapon/weldingtool/WT = P
+				if(!WT.remove_fuel(0, user))
+					user << "The welding tool must be on to complete this task."
+					return
+				playsound(src.loc, 'sound/items/Welder.ogg', 50, 1)
+				if(do_after(user, 20))
+					if(!src || !WT.isOn()) return
+					user << "<span class='notice'>You deconstruct the frame.</span>"
+					new /obj/item/stack/material/steel( src.loc, 5 )
+					qdel(src)
+		if(1)
+			if(istype(P, /obj/item/weapon/wrench))
+				playsound(src.loc, 'sound/items/Ratchet.ogg', 50, 1)
+				if(do_after(user, 20))
+					user << "<span class='notice'>You unfasten the frame.</span>"
+					src.anchored = 0
+					src.state = 0
+			if(istype(P, /obj/item/weapon/circuitboard) && !circuit)
+				var/obj/item/weapon/circuitboard/B = P
+				if(B.board_type == frame_type)
+					playsound(src.loc, 'sound/items/Deconstruct.ogg', 50, 1)
+					user << "<span class='notice'>You place the circuit board inside the frame.</span>"
+					src.icon_state = "[frame_type]_1"
+					src.circuit = P
+					user.drop_item()
+					P.loc = src
+
+					if(frame_type == "machine")  //because machines are assholes
+						components = list()
+						req_components = circuit.req_components.Copy()
+						for(var/A in circuit.req_components)
+							req_components[A] = circuit.req_components[A]
+						req_component_names = circuit.req_components.Copy()
+						for(var/A in req_components)
+							var/cp = text2path(A)
+							var/obj/ct = new cp() // have to quickly instantiate it get name
+							req_component_names[A] = ct.name
+						update_desc()
+						user << desc
```

Comment:
This whole block from line 91-101 looks near identical to 197-207 as far as I can tell. You probably want to break it out into a separate proc and call it from both spots.